app/javascript/article-form/components/TagsField.jsx

```diff
@@ -1,34 +1,85 @@
 import { h } from 'preact';
+import { useEffect, useState } from 'preact/hooks';
 import PropTypes from 'prop-types';
-import { Tags } from '../../shared/components/tags';
+import { TagAutocompleteOption } from './TagAutocompleteOption';
+import { TagAutocompleteSelection } from './TagAutocompleteSelection';
+import { MultiSelectAutocomplete } from '@crayons';
+import { fetchSearch } from '@utilities/search';
 
-export const DEFAULT_TAG_FORMAT = '[0-9A-Za-z, ]+';
+/**
+ * TagsField for the article form. Allows users to search and select up to 4 tags.
+ *
+ * @param {Function} onInput Callback to sync selections to article form state
+ * @param {string} defaultValue Comma separated list of any currently selected tags
+ * @param {Function} switchHelpContext Callback to switch the help context when the field is focused
+ */
+export const TagsField = ({ onInput, defaultValue, switchHelpContext }) => {
+  const [defaultSelections, setDefaultSelections] = useState([]);
+  const [topTags, setTopTags] = useState([]);
+
+  useEffect(() => {
+    window
+      .fetch('/tags/suggest')
+      .then((res) => res.json())
+      .then((results) => setTopTags(results));
+  }, []);
+
+  useEffect(() => {
+    // If the default selections have not already been populated, fetch the tag data
+    if (defaultValue && defaultValue !== '' && defaultSelections.length === 0) {
+      // The article stores tags as a comma separated string
+      const tagNames = defaultValue.split(', ');
+
+      // We need to fetch the full tag data to display the rich UI
+      const tagRequests = tagNames.map((tagName) =>
+        fetchSearch('tags', { name: tagName }).then(
+          ({ result }) => result[0] || { name: tagName },
+        ),
+      );
+
+      Promise.all(tagRequests).then((data) => {
```

Comment:
AFAIK we don't have a bulk fetch endpoint for tag details, so at the minute the tag data for previously added tags are being fetched individually. It's not ideal, but I figure its _OK_ since there is only ever a maximum of 4. Open to other suggestions though

Reply:
I could update `SearchController#tags` for you if you want. Let me know, we can always do this in a follow-up refactoring.

Reply:
What would be awesome is if we had a bulk GET endpoint for tags (like we do for follow statuses, for example) - in this particular case I don't really want to search, as it's only exact matches I'm interested in. If you have time/inclination, it would be awesome, but I could just create a little follow up issue for the pod 🙂

Reply:
Maybe make an issue and label it a quick win, without assigning it to a specific pod, then I can potentially pick it up later this week.

Reply:
Good plan! Added an issue here: https://github.com/forem/forem/issues/16162